lib/ansible/galaxy/dependency_resolution/dataclasses.py

```diff
@@ -441,6 +440,12 @@ def from_requirement_dict(cls, collection_req, art_mgr, validate_signature_optio
 
         if req_type not in {'galaxy', 'subdirs'} and req_name is None:
             req_name = art_mgr.get_direct_collection_fqcn(tmp_inst_req)  # TODO: fix the cache key in artifacts manager?
+        elif req_type != 'galaxy' and req_source is not None and req_name is not None:
+            display.warning(
+                f"A collection name and source were provided for the requirement {collection_req}, "
+                "but only the source will determine the installed collection name."
+            )
+            req_name = None
```

Comment:
Would it make sense to move this to an `elif` branch of the very first `if req_type is None:` block in this classmethod?

Reply:
I was thinking the warning could be meaningful whether or not we have to infer the req_type, so I put it below, but I will take a closer look as I work through the method.